src/orion/core/worker/warm_start/knowledge_base.py

```diff
@@ -0,0 +1,168 @@
+    ):
+        self.storage = storage
+        self.similarity_metric = similarity_metric
+
+    def get_related_trials(
+        self,
+        target_experiment: Experiment | ExperimentClient | ExperimentConfig,
+        max_trials: int | None = None,
+    ) -> list[tuple[ExperimentConfig, list[Trial]]]:
+        """Retrieve the trials from experiments 'similar' to `target_experiment`.
+
+        When `max_trials` is given, only up to `max_trials` are returned in total.
+
+        Parameters
+        ----------
+        target_experiment : Union[Experiment, ExperimentClient, ExperimentConfig]
+            The target experiment, or experiment client, or experiment configuration.
+        max_trials : int, optional
+            Maximum total number of trials to fetch. By default `None`, in which case
+            all trials are returned.
+
+        Returns
+        -------
+        list[tuple[ExperimentConfig, list[Trial]]]
+            A list of tuples of experiments and the similar trials from that experiment.
+        """
+        target_config: ExperimentConfig
+        if isinstance(target_experiment, dict):
+            target_config = target_experiment
+        else:
+            target_config = target_experiment.configuration
+
+        experiment_configs = self.storage.fetch_experiments({})
+        # NOTE: Here we remove the target experiment from the list of experiments.
+        # This occurs because we use a singleton for the storage, so creating the target experiment
+        # implicitly adds it to the storage (and to the knowledge base).
+        experiment_configs = _remove_target_config(target_config, experiment_configs)
+
+        if self.similarity_metric:
+            # Order the experiments with decreasing similarity w.r.t the target experiment.
+            sort_fn = partial(self.similarity_metric, target_config)
+            experiment_configs = sorted(experiment_configs, key=sort_fn, reverse=True)
+
+        return self._get_trials(experiment_configs, max_trials=max_trials)
+
+    @property
+    def n_stored_experiments(self) -> int:
+        """Returns the current number of experiments registered in the Knowledge base.
+
+        NOTE: IF the target experiment is using the same storage object as the KB, then this
+        will also count the target experiment.
+        """
+        return len(self.storage.fetch_experiments({}))
+
+    def _get_trials(
+        self, experiment_configs: Iterable[ExperimentConfig], max_trials: int | None
+    ) -> list[tuple[ExperimentConfig, list[Trial]]]:
+        """Takes at most `max_trials` trials from the given experiments.
+
+        If `max_trials` is None, returns all trials from all given experiments.
+        """
+        related_trials: list[tuple[ExperimentConfig, list[Trial]]] = []
+        total_trials_so_far = 0
+        for experiment_config in experiment_configs:
+            experiment_id = experiment_config["_id"]
+            if experiment_id is None:
```

Comment:
What were these `if` supposed to process? Experiments should always have an ID if they are fetched from the storage.

Reply:
The type of the `_id` field of the Experiment class is currently `str | int | None`. This `if` clause is to take care of that possibility, but you're right, it will never be None in this case. However, type-correctness requires me to add this case here.

Reply:
"fixed" in 8f23392155a38400310c6e6922c7141de4b8652a